File: app/src/main/java/org/magnum/logger/MapActivity.java
package org.magnum.logger;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.FragmentActivity;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnMapClickListener;
import com.google.android.gms.maps.GoogleMapOptions;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.Random;

public class MapActivity extends FragmentActivity implements OnMapClickListener
{
	private static final String TAG = MapActivity.class.getCanonicalName();
	private int numberOfClicks = 0;

	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_map);
		try
		{
			// TODO: Should handle when not connected to the internet
			SupportMapFragment fragment = SupportMapFragment.newInstance(getOptions());
			getSupportFragmentManager().beginTransaction().add(android.R.id.content, fragment, TAG).commit();
			displayDialog(this);

		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}

	}

	private void displayDialog(final OnMapClickListener listener)
	{
		try
		{
			new AlertDialog.Builder(this).setMessage("Please select 3 random points").setPositiveButton("OK", new DialogInterface.OnClickListener()
			{
				public void onClick(DialogInterface dialog, int id)
				{
					((SupportMapFragment) getSupportFragmentManager().findFragmentByTag(TAG)).getMap().setOnMapClickListener(listener);
				}
			}).show();
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}

	}

	private GoogleMapOptions getOptions()
	{
		GoogleMapOptions options = new GoogleMapOptions();
		try
		{
			LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
			if (locationManager != null)
			{
				Location location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
				options.mapType(GoogleMap.MAP_TYPE_NORMAL).camera(new CameraPosition.Builder().target(new LatLng(location.getLatitude(), location.getLongitude())).zoom(10).build()).zoomControlsEnabled(true);
			}
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}
		return options;

	}

	@Override
	public void onBackPressed()
	{
		if (numberOfClicks < 3)
		{
			displayDialog(this);
			return;
		}
		super.onBackPressed();
	}

	@Override
	protected void onPause()
	{
		super.onPause();
		try
		{
			LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
			if (locationManager != null)
			{
				Location last = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
				for (; numberOfClicks < 3; numberOfClicks++)
				{
					// User did not select three points
					// Select random points
					LatLng location = new LatLng(last.getLatitude() + new Random(System.currentTimeMillis()).nextInt(10) - 5, last.getLongitude() + new Random(System.currentTimeMillis()).nextInt(10) - 5);
					setPreferences(location);
				}
			}
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}
	}

	@Override
	public void onMapClick(LatLng location)
	{
		try
		{
			numberOfClicks++;
			((SupportMapFragment) getSupportFragmentManager().findFragmentByTag(TAG)).getMap().addMarker(new MarkerOptions().position(location));
			setPreferences(location);
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}

	}

	private void setPreferences(LatLng location)
	{
		try
		{
			String lng = "Lng" + numberOfClicks;
			String lat = "Lat" + numberOfClicks;
			PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat(lng, (float) location.longitude).commit();
			PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat(lat, (float) location.latitude).commit();
			if (numberOfClicks >= 3)
			{
				Location l3 = new Location(LocationManager.NETWORK_PROVIDER);
				l3.setLatitude(location.latitude);
				l3.setLongitude(location.longitude);
				Location l2 = new Location(LocationManager.NETWORK_PROVIDER);
				l2.setLatitude(PreferenceManager.getDefaultSharedPreferences(this).getFloat("Lat" + 2, 0));
				l2.setLongitude(PreferenceManager.getDefaultSharedPreferences(this).getFloat("Lng" + 2, 0));

				PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat("scale", l3.distanceTo(l2)).commit();

				finish();
				Toast.makeText(this, "Application is now running in background", Toast.LENGTH_LONG).show();
			}
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}

	}
}

File: app/src/main/java/org/magnum/logger/MapActivity.kt
package org.magnum.logger

import android.Manifest
import android.app.AlertDialog
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.fragment.app.FragmentActivity
import androidx.preference.PreferenceManager
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnMapClickListener
import com.google.android.gms.maps.GoogleMapOptions
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import java.util.*

class MapActivity : FragmentActivity(), OnMapClickListener {
    private var numberOfClicks = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_map)
        try {
            val fragment = SupportMapFragment.newInstance(options)
            supportFragmentManager.beginTransaction().add(android.R.id.content, fragment, TAG).commit()
            displayDialog(this)
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    private fun displayDialog(listener: OnMapClickListener) {
        try {
            AlertDialog.Builder(this).setMessage("Please select 3 random points").setPositiveButton("OK") { _, _ ->
                val fragment = mapFragment
                fragment?.getMapAsync { map -> map.setOnMapClickListener(listener) }
            }.show()
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    private val mapFragment get() = supportFragmentManager.findFragmentByTag(TAG) as SupportMapFragment?

    private val options: GoogleMapOptions
        get() {
            val options = GoogleMapOptions()
            try {
                val location = lastKnownLocation
                if (location != null) {
                    options.mapType(GoogleMap.MAP_TYPE_NORMAL).camera(CameraPosition.Builder().target(LatLng(location.latitude, location.longitude)).zoom(10f).build()).zoomControlsEnabled(true)
                }
            } catch (e: Exception) {
                Log.e(TAG, e.toString())
            }
            return options
        }

    override fun onBackPressed() {
        if (numberOfClicks < 3) {
            displayDialog(this)
            return
        }
        super.onBackPressed()
    }

    override fun onPause() {
        super.onPause()
        try {
            val last = lastKnownLocation
            if (last != null) {
                while (numberOfClicks < 3) {

                    // User did not select three points
                    // Select random points
                    val location = LatLng(last.latitude + Random(System.currentTimeMillis()).nextInt(10) - 5, last.longitude + Random(System.currentTimeMillis()).nextInt(10) - 5)
                    setPreferences(location)
                    numberOfClicks++
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    private val lastKnownLocation: Location?
        get() {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), 100)
            } else {
                val locationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
                return locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
            }
            return null
        }

//    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
//        super.onActivityResult(requestCode, resultCode, data)
//        if (requestCode == 100 && resultCode == Activity.RESULT_OK) {
//        }
//    }

    override fun onMapClick(location: LatLng) {
        try {
            numberOfClicks++
            val fragment = supportFragmentManager.findFragmentByTag(TAG) as SupportMapFragment?
            fragment?.getMapAsync { map -> map.addMarker(MarkerOptions().position(location)) }
            setPreferences(location)
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    private fun setPreferences(location: LatLng) {
        try {
            val lng = "Lng$numberOfClicks"
            val lat = "Lat$numberOfClicks"
            PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat(lng, location.longitude.toFloat()).apply()
            PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat(lat, location.latitude.toFloat()).apply()
            if (numberOfClicks >= 3) {
                val l3 = Location(LocationManager.NETWORK_PROVIDER)
                l3.latitude = location.latitude
                l3.longitude = location.longitude
                val l2 = Location(LocationManager.NETWORK_PROVIDER)
                l2.latitude = PreferenceManager.getDefaultSharedPreferences(this).getFloat("Lat" + 2, 0f).toDouble()
                l2.longitude = PreferenceManager.getDefaultSharedPreferences(this).getFloat("Lng" + 2, 0f).toDouble()
                PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat("scale", l3.distanceTo(l2)).apply()
                finish()
                Toast.makeText(this, "Application is now running in background", Toast.LENGTH_LONG).show()
            }
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    companion object {
        private val TAG = MapActivity::class.java.canonicalName
    }
}